### PR TITLE
[IMP] http_client: ApiKey auth

### PR DESCRIPTION
**Pull request by maciej-wichowski**
Sorry for the noise, `black` was triggered by my editor ¯\\_(ツ)_/¯

### CODE REVIEW

File: http_client/models/http_client_auth.py [hunk 1]
```diff
@@ -13,16 +13,17 @@
 from ..value_objects import PathItem
 from .http_client_controller import _raise_endpoint_error
 
-TOKEN_AUTH_PREFIX_MAP = {'bearer': 'Bearer', 'jwt': 'JWT'}
+TOKEN_AUTH_PREFIX_MAP = {"bearer": "Bearer", "jwt": "JWT"}
```
Comment (oerp-odoo) [inline http_client/models/http_client_auth.py]: Please install pre-commit for this repo, so it would run expected rules, instead of your custom ones. It has a bit different pre-commit rules here. Also now its hard to see what was actually changed

And please revert these changes as it should not be part of a change (also `'` quote here is allowed and preferred over `"`)

File: http_client/models/http_client_auth.py [hunk 2]
```diff
@@ -120,85 +122,89 @@ class HttpClientAuth(models.AbstractModel):
...
-        [('draft', "Not Confirmed"), ('confirmed', "Confirmed")],
-        default='draft',
+        [("draft", "Not Confirmed"), ("confirmed", "Confirmed")],
+        default="draft",
         copy=False,
         readonly=True,
         required=True,
     )
+    api_key_header = fields.Char(
+        "API Key header",
```
Comment (oerp-odoo) [inline http_client/models/http_client_auth.py]: We don't really need new field here. We can use `identifier`. Api Key and Basic auth both have key and secret. Its just the difference how they are used (and labels have little meaning behind the scenes).

Of course we can have onchange to auto fill `identifier` with API Key Header default to make it more convenient.

File: http_client/views/http_client_auth_views.xml [hunk 3]
```diff
@@ -78,12 +78,16 @@
...
                         >
-                            <field name="identifier" readonly="state != 'draft'"/>
-                            <field
-                                name="secret"
-                                password="True"
-                                readonly="state != 'draft'"
-                            />
+                            <field name="identifier" readonly="state != 'draft'" invisible="auth_method == 'api_key'"/>
+                            <label for="secret" invisible="auth_method == 'api_key'"/>
+                            <label for="secret" invisible="auth_method != 'api_key'"  string="API Key"/>
```
Comment (oerp-odoo) [inline http_client/views/http_client_auth_views.xml]: Lets not add these workarounds. secret can be used for multiple methods and technically it can be some other kind of "secret", but it is still a secret. Lets keep it simple and have same label (in search and anywhere else, this will still be secret, so it will just add more to confusion). 

Instead I think it would be more neat to have auto generated "description" (possibly HTML), where it would provide you some explanation how to set up each of the auth methods. Let say if you specify `api_key` it would say, you need to use `Identifier` as `Api Key Header` and `Secret` as `Api Key`. 

Then it would be more universal and we would not need to resort to such workarounds:)

Comment (maciej-wichowski) [reply http_client/views/http_client_auth_views.xml]: Sure, I added html description for `api_key` method

File: http_client/models/http_client_auth.py [hunk 4]
```diff
@@ -268,6 +272,9 @@ def get_data(self):
         auth_method = self.auth_method
         if auth_method == 'basic':
             data['auth'] = {'auth': (self.identifier, self.secret)}
+        elif auth_method == 'api_key':
+            header_key = self.identifier or DEFAULT_API_KEY_HEADER
+            data['auth'] = {'headers': {header_key: self.secret}}
```
Comment (oerp-odoo) [inline http_client/models/http_client_auth.py]: We should not fallback to `DEFAULT_API_KEY_HEADER`, because then it would be possible to have empty identifier and silently use default without user realizing it. Instead we have to make `identifier` required on view, to indicate that it needs to be specified when `api_key` method is used.
```suggestion
            data['auth'] = {'headers': {self.identifier: self.secret}}
```

File: http_client/views/http_client_auth_views.xml [hunk 5]
```diff
@@ -86,6 +86,13 @@
                             />
                             <field name="access_token" invisible="1"/>
                             <field name="refresh_token" invisible="1"/>
+                            <div colspan="2">
+                                <p invisible="auth_method != 'api_key'">
```
Comment (oerp-odoo) [inline http_client/views/http_client_auth_views.xml:94]: Will be good enough for now. Later I might implement clear interface for description, so you would provide it in code directly and view would show related description of each method (if it has any).

**Review by oerp-odoo — APPROVED** [review]
There is also `http_client_demo` module, which has most of the tests. It would be nice to have some tests, that actually validate this auth method is working

**Comment by maciej-wichowski** [comment]
@oerp-odoo please let me know if any more tests are needed or merge :)

**Comment by oerp-odoo** [comment]
> @oerp-odoo please let me know if any more tests are needed or merge :)

Looks good, thanks:)
